backends/redis.go

```diff
@@ -55,99 +55,84 @@ func (e KeyError) Error() string {
-		ReadTimeout:      timeout,
-		WriteTimeout:     timeout,
-		DialTimeout:      timeout,
-		TLSConfig:        tlsConfig,
+	conf := r.config
+	optsR := model.RedisOptions{
+		Username:         conf.Username,
+		Password:         conf.Password,
+		Host:             conf.Host,
+		Port:             conf.Port,
+		Timeout:          conf.Timeout,
+		Hosts:            conf.Hosts,
+		Addrs:            conf.Addrs,
+		MasterName:       conf.MasterName,
+		SentinelPassword: conf.SentinelPassword,
+		Database:         conf.Database,
+		MaxActive:        conf.MaxActive,
+		EnableCluster:    conf.EnableCluster,
+	}
+
+	tls := model.TLS{
+		Enable:             conf.UseSSL,
+		InsecureSkipVerify: conf.SSLInsecureSkipVerify,
+		CAFile:             conf.TLS.CAFile,
+		CertFile:           conf.TLS.CertFile,
+		KeyFile:            conf.TLS.KeyFile,
+		MaxVersion:         string(conf.TLS.MaxVersion),
+		MinVersion:         string(conf.TLS.MinVersion),
+	}
+
+	connector, err := connector.NewConnector(model.RedisV9Type, model.WithRedisConfig(&optsR), model.WithTLS(&tls))
+	if err != nil {
+		redisLogger.WithError(err).Error("creating redis connector")
+		return err
 	}
 
-	if opts.MasterName != "" {
-		redisLogger.Info("Creating sentinel-backed fail-over client")
-		client = redis.NewFailoverClient(opts.Failover())
-	} else if r.config.EnableCluster {
-		redisLogger.Info("Creating cluster client")
-		client = redis.NewClusterClient(opts.Cluster())
-	} else {
-		redisLogger.Info("Creating single-node client")
-		client = redis.NewClient(opts.Simple())
+	r.kv, err = temporal.NewKeyValue(connector)
+	if err != nil {
+		redisLogger.WithError(err).Error("creating KV store")
+		return err
 	}
 
-	return client
+	return nil
 }
 
-func (r *RedisBackend) getRedisAddrs() (addrs []string) {
-	if len(r.config.Addrs) != 0 {
-		addrs = r.config.Addrs
-	} else {
-		for h, p := range r.config.Hosts {
-			addr := h + ":" + p
-			addrs = append(addrs, addr)
-		}
-	}
-
-	if len(addrs) == 0 && r.config.Port != 0 {
-		addr := r.config.Host + ":" + strconv.Itoa(r.config.Port)
-		addrs = append(addrs, addr)
+// Init will create the initial in-memory store structures
+func (r *RedisBackend) Init(config interface{}) error {
+	asJ, err := json.Marshal(config)
+	if err != nil {
+		return err
 	}
 
-	return addrs
-}
-
-func (r *RedisBackend) Connect() bool {
-
-	r.db = r.newRedisClusterPool()
-	return true
-}
-
-// Init will create the initial in-memory store structures
-func (r *RedisBackend) Init(config interface{}) {
-	asJ, _ := json.Marshal(config)
 	fixedConf := RedisConfig{}
-	json.Unmarshal(asJ, &fixedConf)
+	err = json.Unmarshal(asJ, &fixedConf)
+	if err != nil {
+		return err
+	}
 	r.config = &fixedConf
-	r.Connect()
+	err = r.Connect()
+	if err != nil {
+		return err
+	}
+
 	redisLogger.Info("Initialized")
+	return nil
 }
 
 // SetDb from existent connection
-func (r *RedisBackend) SetDb(db redis.UniversalClient) {
+func (r *RedisBackend) SetDb(kv temporal.KeyValue) {
```

Comment:
I think that ideally, this method should receive a Connection instead of a KV. What do you think?

Reply:
What would be the benefit? If I receive the connector then anyway I will need to instance my KVStore like:


```
r.kv, err = temporal.NewKeyValue(connector)
	if err != nil {
		redisLogger.WithError(err).Error("creating KV store")
		return err
	}
```